[Mantis 15366] Test: Reliably calc. time remaining

Rillke opened the pull request:
![image](https://user-images.githubusercontent.com/2311611/89765797-52d9bc80-daf7-11ea-9354-9106ab4ccd8b.png)


- The implementation before this fix used an incremented counter based
  approach for calculating time remaining: Every 1000ms add one to
  your counter and assume the counter is the time passed since page
  load in s.
- Modern browsers de-prioritize inactive tabs in order to safe CPU
  and energy.
- This causes `setInterval()` fire less frequently than expected.
- The clock is slow.
- Instead, we use the browser's performance API to calculate the time
  passed since test start. The performance API is supported by all
  modern browsers and does not suffer from above issues, not even from
  time zone or clock changes.
  https://developer.mozilla.org/en-US/docs/Web/API/Performance/now
  https://www.w3.org/TR/hr-time-2/#dom-performance-now
- This is similar to how frames in (3D) games are calculated.
- Additional changes:
    - prevent pollution of global scope wrapping the code in an IIFE
    - begin sort the code units into functions
    - use more strict comparison operators
    - make jshint happy declaring global dependencies

Rillke commented:
Should I undo the white space changes in `class.ilTestPlayerAbstractGUI.php`?

Rillke commented:
Additional note about why we use an offset to an initial `performance.now()`:

> The current high resolution time is the high resolution time from the time origin to the present time (typically called "now").

> The time origin is the time value from which time is measured:
>    If the global object is a Window object, the time origin MUST be equal to:
>        the time when the browsing context is first created if there is no previous document;
>        otherwise, the time of the user confirming the navigation during the previous document's prompt to unload algorithm, if a previous document exists and if the confirmation dialog was displayed;
>        otherwise, the time of starting the navigation responsible for loading the Window object's newest Document object.

The issue here is that the performance timer begins when navigation starts, i.e. when the request is sent to the server, not when the server calculates all the time values filled into this template script. When the server is very busy, or the request is transmitted slowly, we would assume too much time to be passed. That's an issue because auto-save/-submission, which is triggered when the time reaches the test's end, would be before the test actually ends. ILIAS will just display "autosave successful" in this case instead of the expected "The test ended and your last solution has been stored and you will be redirected in a few moments"

mjansenDatabay commented:
@Rillke Thx for your contribution. Indeed, the whitespace changes are a bit distracting when reviewing the changes :-).

@mbecker-databay Please review this PR and merge it accordingly (to other stable branches as well) if you don't have any objections.

Rillke committed:
[Mantis 15366] Test: Reliably calc. time remaining

- The implementation before this fix used a counter based approach
  for calculating time remaining: Every 1000ms add one to your
  counter and assume the counter is the time passed since page
  load.
- Modern browsers de-prioritize inactive tabs in order to save on CPU
  and energy. This affects Firefox, Safari and all Chromium-core
  browsers.
- This causes `setInterval()` fire less frequently than expected.
- The clock is slow and confuses those who take the test.
- The solution:
  Instead, we use the browser's performance API to calculate the time
  passed since test start. The performance API is supported by all
  modern browsers and does not suffer from above issues, not even
  from time zone or clock changes.
  https://developer.mozilla.org/en-US/docs/Web/API/Performance/now
  https://www.w3.org/TR/hr-time-2/#dom-performance-now
- This is similar to how frames in (3D) games are calculated:
  Do not assume a fixed frame rate but use time diff to calculate
  the next frame.
- Additional changes:
    - prevent pollution of global scope wrapping the code in an IIFE
    - sort the code units into functions
    - use more strict comparison operators
    - make jshint happy declaring global dependencies
    - add jsduck-style comments to documenting functions, params
      and return values

Rillke commented:
> [Required and Supported Software ILIAS 6](https://docu.ilias.de/goto_docu_wiki_wpage_5724_1357.html): ILIAS supports current versions of Firefox and Chrome. The browser support for particular views or features can be limited due to browser CSS/JavaScript limitations.

Is this statement final?

Considering [the compatibility matrix](https://caniuse.com/#feat=mdn-api_performance_now), means we can drop the counter-based approach altogether. I am asking because in the past I noticed some commits restoring compatibility in ILIAS with older software.

Rillke commented:
@mjansenDatabay The white space changes complicating review have been removed. Thanks for your quick response.

gvollbach reviewed (approved):
Looks good, thanks for the contribution @Rillke 

@mbecker-databay Review is done.
